[libpas] PGM does not respect alignment

Constellation opened the pull request:
#### 06235d829da16c4218a606419b3044ab2b11adf2
<pre>
[libpas] PGM does not respect alignment
<a href="https://bugs.webkit.org/show_bug.cgi?id=286708">https://bugs.webkit.org/show_bug.cgi?id=286708</a>
<a href="https://rdar.apple.com/143847878">rdar://143847878</a>

Reviewed by Mark Lam.

PGM code is ignoring user-specified alignment. This does not work at all
since the user expects aligned allocations but PGM returns unrelated
pointers. This patch wires alignment correctly to PGM and fixes address
computation.

We also fixed PGMTests.cpp as it cannot be compiled. Also fixes so many
style issues around PGM code.

* Source/bmalloc/libpas/src/libpas/pas_enumerate_large_heaps.c:
(pas_hash_map_entry_callback):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate):
(pas_probabilistic_guard_malloc_deallocate):
(pas_probabilistic_guard_malloc_check_exists):
(pas_probabilistic_guard_malloc_return_as_large_map_entry):
(pas_probabilistic_guard_malloc_debug_info):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(memory_reader_adapter):
(setup_memory_reader):
(pas_update_report_crash_fields):
(pas_report_crash_extract_pgm_failure):
* Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_common.h:
(pas_try_allocate_pgm):
(pas_try_allocate_common_impl_slow):
* Source/bmalloc/libpas/src/libpas/pas_try_reallocate.h:
(pas_try_reallocate):
* Source/bmalloc/libpas/src/test/EnumerationTests.cpp:
(std::testPGMEnumerationBasic):
(std::testPGMEnumerationAddAndFree):
* Source/bmalloc/libpas/src/test/PGMTests.cpp:
(std::testPGMSingleAlloc):
(std::testPGMMultipleAlloc):
(std::testPGMAlignedAlloc):
(std::testPGMRealloc):
(std::testPGMMetaData):
(std::testPGMErrors):
(std::testPGMMetadataVectorManagement):
(std::testPGMMetadataVectorManagementFewDeallocations):
(std::testPGMMetadataDoubleFreeBehavior):
(std::testPGMMetadataVectorManagementRehash):
(addPGMTests):

Canonical link: <a href="https://commits.webkit.org/289637@main">https://commits.webkit.org/289637@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d204cab57da7485c975b68d5920996efeeee88b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87467 "58 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6980 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41840 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92329 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38206 "Built successfully") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89518 "Build is in progress. Recent messages:") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7362 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15147 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67556 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25295 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90469 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5604 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79143 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47900 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5392 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33537 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37322 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/80262 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75810 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34400 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94214 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/86241 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14632 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10733 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76383 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14886 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74998 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75607 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19981 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18398 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7570 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13642 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14650 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/108733 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14393 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26151 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17837 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16176 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/bmalloc/libpas/src/test/PGMTests.cpp:
```diff
@@ -108,17 +109,52 @@ void testPGMMultipleAlloc() {
+        mem_storage[i] = mem;
+        memset(reinterpret_cast<void *>(mem.begin), 0x42, alloc_size);
+        CHECK(pas_is_aligned(mem.begin, alignment));
     }
```
We're only testing with 1 alignment value here.  Can we test this with different alignment values from 4K to 512K?  For the larger alignments, maybe we have to reduce `num_allocations`?

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -93,20 +92,20 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
-
-    size_t mem_to_waste = (page_size - (size % page_size)) % page_size;
+    size_t size_with_alignment_reservation = size + alignment - 1;
+    size_t mem_to_waste = (page_size - (size_with_alignment_reservation % page_size)) % page_size;
```
I think changing the definition of `mem_to_waste` here is going to break `pas_report_crash_extract_pgm_failure()`

Also, for the case of `alignment > page_size` it seems weird to consider only the wasted memory within a page since there will be multiple of page-sized wasted memory (to align the allocation), when it comes to bookkeeping `free_wasted_mem`. And really, there is no wasted space within a page in this case since `size` should be a multiple of `alignment`, yet `mem_to_waste` will be 1 to compensate for the -1 in `size_with_alignment_reservation`. (I'm pretty sure the calling code sanitizes `size` to be a multiple of requested `alignment`.)

So, I wonder if things would be cleaner if we just treat the two cases separately:
1. alignment <= page_size: in this case, `mem_to_waste` is the intra-page wasted memory
2. alignment > page_size: in this case, `mem_to_waste` will always be a multiple of `page_size`, i.e. we need to slide to the correct page to align to.

dhecht replied:
Was talking to Mark about this and we also realized that in the case where alignment > page_size, there's a couple of other differences:
1. We could/should mprotect/madvise the alignment pages before and after the data page rather than just the guard page
2. Two guard pages separate from the alignment pages aren't necessary since alignment pages can do double duty.

In other words, for alignment > page_size case, maybe we should do:

num_data_pages = size / page_size;
num_alignment_pages = alignment / page_size;

Then, allocate num_data_pages + num_alignment_pages + 1 pages, slide the allocation to be aligned and not using the 0th page, and then mprotect and madvise all the pages that aren't part of the user's allocation (guaranteed to have at least 1 guard page before and after the user's allocation).

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -151,18 +150,25 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
+        /* If alignment is large, then there is no enough mem_to_waste to slide. */
+        uintptr_t right_key = pas_round_up(result.begin + page_size + mem_to_waste, alignment);
+        if ((right_key + size + page_size) <= (result.begin + mem_to_alloc)) {
+            right_align = 1;
```
I think we can never get here with `alignment > page_size`, right? In that case, there is only one location we can put the allocation, which is already calculated by `key`, since in that case we've allocated just enough pages to ensure one of them is aligned. So again, I think it might be simpler to treat the two cases separately (alignment vs page_size cases).

Put another way, right vs left alignment is meaningless when `alignment > page_size`.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -81,80 +80,113 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
+         * Also, to achieve this alignment request, we need this expanded size, and this is not something wasted for PGM.
+         * So we set 0 to mem_to_waste.
+         */
+        size_t size_with_alignment_reservation = pas_round_up(size + alignment - 1, page_size);
```
hmm, why this -1 here? Doesn't `pas_round_up` already that?
e.g. suppose size=16k+1, alignment=8k. With this -1, we get the same answer as when size=16k, which doesn't make sense.
I think it practice it doesn't matter assuming the caller already rounded up size to alignment, but this still seems weird.
Maybe the intent was to subtract `page_size` since that's effectively the unit of addressing on this path, and then adding the 2 pages in below would be necessary.

Constellation replied:
I think `size + alignment - 1` is right. The worst case of allocation is the obtained pointer is `alignment - 1`, in that case, we need `alignment - 1` shift of the pointer, so the necessary size to get aligned region is always `size + alignement - 1`.

Constellation replied:
Let's say allocated thing is 1 byte ahead of the alignment boundary, then we need alignment - 1 shift to get the next boundary.

Constellation replied:
For example, if the size is much smaller than the alignment, like, `size = 16` while `alignment = 16KB`, then we at least need to allocate `16KB + 15` to find 16KB boundary.

Constellation replied:
But since the unit is one-page, we can consider the worst case scenario, like, alignment-boundary + one-page as the result pointer. In that case, we need to add alignment - one-page to shift to the next alignment boundary. So the size can be `pas_round_up(size + alignment - page_size, page_size);`

dhecht replied:
Right, I was using the fact that the underlying allocation is done with `page_size` alignment, which then in my mind makes it easier to think about the allocation units as `page_size`:
`result = pas_large_heap_try_allocate_and_forget(large_heap, mem_to_alloc, page_size, allocation_mode, heap_config, transaction);`

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -81,80 +80,113 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
+         */
+        size_t size_with_alignment_reservation = pas_round_up(size + alignment - 1, page_size);
+        mem_to_waste = 0;
+        mem_to_alloc = (2 * page_size) + size_with_alignment_reservation;
```
I think we only need to add one additional page since the alignment reservation will always result in at least one page of padding on either the left or the right (or both), so we just need one additional page to guarantee we can have guards on both sides.

Constellation replied:
`size_with_alignment_reservation` only ensures that alignment boundary can exist in the region and we can have size after that. This means that it can be the last part of the region, and in that case we do not have the next page. So I think 2 pages are necessary here.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -81,80 +80,116 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
+         * mem_to_waste is also x * alignment. Thus right-align will be guaranteed to be aligned.
+         */
+        size_t size_with_alignment_reservation = pas_round_up(size, alignment);
+        mem_to_waste = (page_size - (size_with_alignment_reservation % page_size)) % page_size;
```
Recording my reasoning here for posterity: The only time `% page_size` is interesting is if `(size_with_alignment_reservation % page_size)` == `page_size`.   Though we know that `alignment < page_size`, we don't know what `size` is. Hence, `size_with_alignment_reservation` can be exactly an even number of page, which leads:
    `(size_with_alignment_reservation % page_size)` == `0`
    ==> `(page_size - (size_with_alignment_reservation % page_size))` == `page_size`

... the waste here is 0, not `page_size`.  Hence, the `% page_size` is needed.

MenloDorian commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -81,80 +80,116 @@ pas_allocation_result pas_probabilistic_guard_malloc_allocate(pas_large_heap* la
+         */
+        size_t size_with_alignment_reservation = pas_round_up(size + alignment - page_size, page_size);
+        mem_to_waste = 0;
+        mem_to_alloc = (2 * page_size) + size_with_alignment_reservation;
```
Recording my reasoning here for posterity: the allocated memory pages might be already aligned at the start.  This means that we'll have to return the data portion at `start` + `alignment`.  That means `mem_to_alloc` must be `alignment` (can fit leading guard page) + `size` + `page_size` (for trailing guard page).

`alignment` + `size` + `page_size`
== (`(size + alignment - page_size)` + `page_size`) + `page_size`
== `size_with_alignment_reservation` + `(2 * page_size)`

Hand waving a bit here, but `(size + alignment - page_size)` effectively == `size_with_alignment_reservation` after we consider needed rounding.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[libpas] PGM does not respect alignment
https://bugs.webkit.org/show_bug.cgi?id=286708
rdar://143847878

Reviewed by Mark Lam.

PGM code is ignoring user-specified alignment. This does not work at all
since the user expects aligned allocations but PGM returns unrelated
pointers. This patch wires alignment correctly to PGM and fixes address
computation.

We also fixed PGMTests.cpp as it cannot be compiled. Also fixes so many
style issues around PGM code.

* Source/bmalloc/libpas/src/libpas/pas_enumerate_large_heaps.c:
(pas_hash_map_entry_callback):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_allocate):
(pas_probabilistic_guard_malloc_deallocate):
(pas_probabilistic_guard_malloc_check_exists):
(pas_probabilistic_guard_malloc_return_as_large_map_entry):
(pas_probabilistic_guard_malloc_debug_info):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
* Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
(memory_reader_adapter):
(setup_memory_reader):
(pas_update_report_crash_fields):
(pas_report_crash_extract_pgm_failure):
* Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
* Source/bmalloc/libpas/src/libpas/pas_try_allocate_common.h:
(pas_try_allocate_pgm):
(pas_try_allocate_common_impl_slow):
* Source/bmalloc/libpas/src/libpas/pas_try_reallocate.h:
(pas_try_reallocate):
* Source/bmalloc/libpas/src/test/EnumerationTests.cpp:
(std::testPGMEnumerationBasic):
(std::testPGMEnumerationAddAndFree):
* Source/bmalloc/libpas/src/test/PGMTests.cpp:
(std::testPGMSingleAlloc):
(std::testPGMMultipleAlloc):
(std::testPGMAlignedAlloc):
(std::testPGMRealloc):
(std::testPGMMetaData):
(std::testPGMErrors):
(std::testPGMMetadataVectorManagement):
(std::testPGMMetadataVectorManagementFewDeallocations):
(std::testPGMMetadataDoubleFreeBehavior):
(std::testPGMMetadataVectorManagementRehash):
(addPGMTests):

Canonical link: https://commits.webkit.org/289637@main
